src/cluster.c

```diff
@@ -5146,6 +5147,11 @@ void clusterSetMaster(clusterNode *n) {
     serverAssert(n != myself);
     serverAssert(myself->numslots == 0);
 
+    /* If the node is still in the same shard, don't need to unsubscribe the clients. */
+    if (memcmp(myself->shard_id, n->shard_id, CLUSTER_NAMELEN) != 0) {
+            removeAllShardChannelSubscription();
```

Comment:
```suggestion
        removeAllShardChannelSubscription();
```

Reply:
The only challenge which we discussed earlier was that if it stays within the same shard it would lead to a few unnecessary client reconnection.

https://github.com/redis/redis/pull/12577#discussion_r1330565425